ext/dom/php_dom.c

```diff
@@ -1569,14 +1924,60 @@ void dom_reconcile_ns_list(xmlDocPtr doc, xmlNodePtr nodep, xmlNodePtr last)
 	 * dom_libxml_reconcile_ensure_namespaces_are_declared() performs its own recursion. */
 	while (true) {
 		/* The internal libxml2 call will already check the node type, no need for us to do it here. */
-		dom_libxml_reconcile_ensure_namespaces_are_declared(nodep);
+		xmlReconciliateNs(doc, nodep);
 		if (nodep == last) {
 			break;
 		}
 		nodep = nodep->next;
 	}
 }
 
+/* https://dom.spec.whatwg.org/#validate-and-extract */
+int dom_validate_and_extract(const zend_string *namespace, const zend_string *qname, xmlChar **localName, xmlChar **prefix)
+{
+	/* 1. If namespace is the empty string, then set it to null.
+	 *    However, we're going to cheat and do the opposite to make
+	 *    implementation of the below steps with existing zend_ helpers easier. */
+	if (namespace == NULL) {
+		namespace = zend_empty_string;
+	}
+
+	/* 2. Validate qualifiedName. */
+	if (xmlValidateQName(BAD_CAST ZSTR_VAL(qname), /* allow spaces */ 0) != 0) {
+		return INVALID_CHARACTER_ERR;
+	}
+
+	/* Steps 3-5 */
+	*localName = xmlSplitQName2(BAD_CAST ZSTR_VAL(qname), prefix);
+
+	/* 6. If prefix is non-null and namespace is null, then throw a "NamespaceError" DOMException.
+	 *    Note that null namespace means empty string here becaue of step 1. */
+	if (*prefix != NULL && ZSTR_VAL(namespace)[0] == '\0') {
```

Comment:
Can't you check the length of the namespace string here?

Reply:
Yes; although that allows passing "\0" which will cause strange results (embedded NUL is not valid for namespace uris), hence the reason I did the check like this.

Reply:
Shouldn't a string that does not have nul bytes be checked in advance?

Reply:
Maybe... Although this is never done in DOM atm. I can make this change for new DOM, replacing a lot of s & S with p & P in the argument parsing...

Reply:
Right, let's leave this as a follow-up then